tutorials/shaders/shader_reference/shader_preprocessor.rst

```diff
@@ -1,56 +1,251 @@
+    #define USE_LIGHT 0 // Evaluates to `false`.
+    #define USE_COLOR 1 // Evaluates to `true`.
 
     #if VAR == 3 && (USE_LIGHT || USE_COLOR)
+    // Condition is `true`. Include this portion in the final shader.
+    #endif
+
+Using the ``defined()`` *preprocessor function*, you can check whether the
+passed identifier is defined a by ``#define`` placed above that directive. This
+is useful for creating multiple shader versions in the same file. It may be
+continued by a `#else` block, but must be ended with the ``#endif`` directive.
+
+The ``defined()`` function's result can be negated by using the ``!`` (boolean NOT)
+symbol in front of it. This can be used to check whether a define is *not* set.
+
+.. code-block:: glsl
+
+    #define USE_LIGHT
+    #define USE_COLOR
+
+    // Correct syntax:
+    #if defined(USE_LIGHT) || defined(USE_COLOR) || !defined(USE_REFRACTION)
+    // Condition is `true`. Include this portion in the final shader.
+    #endif
+
+Be careful, as ``defined()`` must only wrap a single identifier within parentheses, never more:
+
+.. code-block:: glsl
+
+    // Incorrect syntax (parentheses are not placed where they should be):
+    #if defined(USE_LIGHT || USE_COLOR || !USE_REFRACTION)
+    // This will cause an error or not behave as expected.
+    #endif
 
+.. tip::
 
+    In the shader editor, preprocessor branches that evaluate to ``false`` (and
+    are therefore excluded from the final compiled shader) will appear grayed
+    out. This does not apply to run-time ``if`` statements.
+
+**#if preprocessor versus if statement: Performance caveats**
+
+The :ref:`shading language <doc_shading_language>` supports run-time ``if`` statements:
+
+.. code-block:: glsl
+
+    uniform bool USE_LIGHT = true;
+
+    if (USE_LIGHT) {
+        // This part is included in the compiled shader, and always run.
+    } else {
+        // This part is included in the compiled shader, but never run.
+    }
+
+If the uniform is never changed, this behaves identical to the following usage
+of the ``#if`` preprocessor statement:
+
+.. code-block:: glsl
+
+    #define USE_LIGHT
+
+    #if defined(USE_LIGHT)
+    // This part is included in the compiled shader, and always run.
+    #else
+    // This part is *not* included in the compiled shader (and therefore never run).
+    #endif
+
+However, the ``#if`` variant can be faster in certain scenarios. This is because
+all run-time branches in a shader are still compiled and take up register space,
+even if they are never run in practice.
+
+`Modern GPUs are quite effective at performing "static" branching <https://medium.com/@jasonbooth_86226/branching-on-a-gpu-18bfc83694f2>`__.
+"Static" branching refers to ``if`` statements where *all* pixels/vertices
+evaluate to ``true`` or ``false`` in a given frame. However, high amounts of
+:abbr:`VGPR (Vector General-Purpose Register)`s (which can be caused by having
+too many branches) can still slow down shader execution significantly.
```

Comment:
```suggestion
evaluate to the same result in a given shader invocation. However, high amounts of
:abbr:`VGPR (Vector General-Purpose Register)`s (which can be caused by having
too many branches) can still slow down shader execution significantly.
```

This might be too much information for the manual. It is incorrect to say "frame", but saying wave or warp is likely meaningless to the type of reader that this document is geared towards. I went with shader invocation which is closer to true